src/http/http2.h

```diff
@@ -0,0 +1,604 @@
+    nghttp2_session* session, const nghttp2_frame* frame, void* user_data)
+  {
+    LOG_TRACE_FMT(
+      "http2::on_frame_recv_callback_client, type: {}", frame->hd.type);
+
+    auto* s = get_session(user_data);
+    auto* stream_data = get_stream_data(session, frame->hd.stream_id);
+
+    switch (frame->hd.type)
+    {
+      case NGHTTP2_DATA:
+      {
+        LOG_TRACE_FMT("Data frame: {}", frame->headers.cat);
+        if (stream_data->id == frame->hd.stream_id)
+        {
+          LOG_DEBUG_FMT("All headers received");
+          s->handle_response(stream_data);
+        }
+        break;
+      }
+      default:
+      {
+        break;
+      }
+    }
+    return 0;
+  }
+
+  static int on_begin_headers_callback(
+    nghttp2_session* session, const nghttp2_frame* frame, void* user_data)
+  {
+    LOG_TRACE_FMT("http2::on_begin_headers_callback");
+
+    auto* s = get_session(user_data);
+    auto stream_data = std::make_shared<StreamData>(frame->hd.stream_id);
+    s->add_stream(stream_data);
+    auto rc = nghttp2_session_set_stream_user_data(
+      session, frame->hd.stream_id, stream_data.get());
+    if (rc != 0)
+    {
+      throw std::logic_error(fmt::format(
+        "HTTP/2: Could not set user data for stream {}: {}",
+        frame->hd.stream_id,
+        nghttp2_strerror(rc)));
+    }
+
+    return 0;
+  }
+
+  static int on_begin_headers_callback_client(
+    nghttp2_session* session, const nghttp2_frame* frame, void* user_data)
+  {
+    LOG_TRACE_FMT(
+      "http2::on_begin_headers_callback_client: {}", frame->hd.type);
+
+    return 0;
+  }
+
+  static int on_header_callback(
+    nghttp2_session* session,
+    const nghttp2_frame* frame,
+    const uint8_t* name,
+    size_t namelen,
+    const uint8_t* value,
+    size_t valuelen,
+    uint8_t flags,
+    void* user_data)
+  {
+    auto k = std::string(name, name + namelen);
+    auto v = std::string(value, value + valuelen);
+    LOG_TRACE_FMT("http2::on_header_callback: {}:{}", k, v);
+
+    auto* s = get_session(user_data);
+    auto* stream_data = get_stream_data(session, frame->hd.stream_id);
+
+    if (k == http2::headers::PATH)
+    {
+      stream_data->url = v;
+    }
+    else if (k == http2::headers::METHOD)
+    {
+      stream_data->verb = v;
+    }
+    else
+    {
+      stream_data->headers.emplace(k, v);
+    }
+
+    return 0;
+  }
+
+  static int on_header_callback_client(
+    nghttp2_session* session,
+    const nghttp2_frame* frame,
+    const uint8_t* name,
+    size_t namelen,
+    const uint8_t* value,
+    size_t valuelen,
+    uint8_t flags,
+    void* user_data)
+  {
+    auto k = std::string(name, name + namelen);
+    auto v = std::string(value, value + valuelen);
+    LOG_TRACE_FMT("http2::on_header_callback_client: {}:{}", k, v);
+
+    auto* stream_data = get_stream_data(session, frame->hd.stream_id);
+
+    if (k == http2::headers::STATUS)
+    {
+      // Note: handle errors
+      stream_data->status = HTTP_STATUS_OK;
```

Comment:
Doesn't this mean that HTTP2 client requests can never fail? It's probably easy to add by just parsing `v`.